diff --git a/dwm.c b/dwm.c
index 5e4d494..e482dc7 100644
--- a/dwm.c
+++ b/dwm.c
@@ -182,6 +182,7 @@ static void manage(Window w, XWindowAttributes *wa);
 static void mappingnotify(XEvent *e);
 static void maprequest(XEvent *e);
 static void monocle(Monitor *m);
+static void grid(Monitor *m);
 static void motionnotify(XEvent *e);
 static void movemouse(const Arg *arg);
 static Client *nexttiled(Client *c);
@@ -1116,6 +1117,34 @@ monocle(Monitor *m)
 		resize(c, m->wx, m->wy, m->ww - 2 * c->bw, m->wh - 2 * c->bw, 0);
 }
 
+void
+grid(Monitor *m) {
+        unsigned int i, n, cx, cy, cw, ch, aw, ah, cols, rows;
+        Client *c;
+
+        for(n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next))
+                n++;
+
+        /* grid dimensions */
+        for(rows = 0; rows <= n/2; rows++)
+                if(rows*rows >= n)
+                        break;
+        cols = (rows && (rows - 1) * rows >= n) ? rows - 1 : rows;
+
+        /* window geoms (cell height/width) */
+        ch = m->wh / (rows ? rows : 1);
+        cw = m->ww / (cols ? cols : 1);
+        for(i = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next)) {
+                cx = m->wx + (i / rows) * cw;
+                cy = m->wy + (i % rows) * ch;
+                /* adjust height/width of last row/column's windows */
+                ah = ((i + 1) % rows == 0) ? m->wh - ch * rows : 0;
+                aw = (i >= rows * (cols - 1)) ? m->ww - cw * cols : 0;
+                resize(c, cx, cy, cw - 2 * c->bw + aw, ch - 2 * c->bw + ah, False);
+                i++;
+        }
+}
+
 void
 motionnotify(XEvent *e)
 {
